add fltk example

MoAlyousef opened the pull request:
Hello
Thank you for the nice and easy to use crate. This PR adds an example using the fltk crate. The window passed to Pixels can be the top level window (like in the example) or an embedded window. I took the liberty to change the minimal example to be of a circle bouncing around.
Thanks again

bjorn3 commented on examples/minimal-fltk/Cargo.toml:
```diff
@@ -0,0 +1,16 @@
+fltk = "0.14"
+env_logger = "0.7.1"
+log = "0.4.8"
+pixels = { path = "../.." }
```
nit: missing trailing newline

MoAlyousef replied:
oh. I didn't know a new line was required.

fixed

parasyte commented:
This looks good to me at first glance. @MoAlyousef please rebase or merge the latest HEAD to pick up an unrelated CI patch.

MoAlyousef commented:
> This looks good to me at first glance. @MoAlyousef please rebase or merge the latest HEAD to pick up an unrelated CI patch.

Sure

parasyte commented:
Ok, I tested this on macOS (11.2, Big Sur) and I don't get anything drawn on the fltk window. I also tested the `pong` demo from the fltk repo, and it does work. So there seems to be something missing. Debugging reveals that the closure passed to `win.draw()` does not run at all. Any thoughts?

Also tried this PR on Windows and it works fine there.

MoAlyousef commented:
Unfortunately I can't get any of the other Pixels demos/examples to run on my MacOS Big Sur (installed on a macbook air 2015). 
Even with the PIXELS_HIGH_PERF env variable set prior to running the build, so I assumed it's just my drivers. Normally however calls to win.redraw() should call the win.draw() callback.

MoAlyousef commented:
Anyways the win.draw() call is not really necessary. This for example works (on my windows machine):
```rust
fn main() -> Result<(), Error> {
    env_logger::init();
    let app = app::App::default();
    let mut win = Window::default()
        .with_size(WIDTH as i32, HEIGHT as i32)
        .with_label("Hello Pixels");
    win.end();
    win.make_resizable(true);
    win.show();

    let mut pixels = {
        let surface_texture = SurfaceTexture::new(WIDTH, HEIGHT, &win);
        Pixels::new(WIDTH, HEIGHT, surface_texture)?
    };

    let mut world = World::new();

    while app.wait() {
        // Update internal state
        world.update();
        // Draw the current frame
        world.draw(pixels.get_frame());
        pixels.render().unwrap()
        app::redraw();
        // Calls to redraw in the event loop require an explicit sleep
        thread::sleep(Duration::from_millis(SLEEP));
    }

    Ok(())
}
```

parasyte commented:
> Unfortunately I can't get any of the other Pixels demos/examples to run on my MacOS Big Sur (installed on a macbook air 2015).
> Even with the PIXELS_HIGH_PERF env variable set prior to running the build, so I assumed it's just my drivers. Normally however calls to win.redraw() should call the win.draw() callback.
> 
> Edit: I'll try to investigate things further on a colleague's device.

Thanks. Let me know if there is anything I can help with on the `pixels` side. I do want to make sure the crates works on all modern hardware, but macOS tends to be tricky... #72 #76 #87 and others...

> Anyways the win.draw() call is not really necessary. This for example works (on my windows machine):

Awesome! This gets it a little farther along on macOS. I had to change `app::redraw()` back to `win.redraw()`. The former only causes the window to redraw when the window receives events (like moving the mouse cursor).

MoAlyousef commented:
I've updated the example and removed the call to win.draw(). FLTK documents that calling Fl_Widget::draw on apple systems are expected to call [Quartz drawing](https://www.fltk.org/doc-1.3/osissues.html#osissues_macos) functions (or those wrapped by FLTK itself). Apparently the code gets optimized out since there are no draw calls. 
`All code inside Fl_Widget::draw() is expected to call Quartz drawing functions.` 
Anyways, I also tried some of the wgpu examples and all give me blank windows on my macbook air. I still haven't had the chance to try things out on my friend's Mac.

parasyte commented:
Thank you for your contribution! The addition of an example of `fltk` provides more options for developers to use GPU-acceleration in their applications. I appreciate it.

> FLTK documents that calling Fl_Widget::draw on apple systems are expected to call [Quartz drawing](https://www.fltk.org/doc-1.3/osissues.html#osissues_macos) functions (or those wrapped by FLTK itself). Apparently the code gets optimized out since there are no draw calls.
> `All code inside Fl_Widget::draw() is expected to call Quartz drawing functions.`

This is a bit surprising. I'm glad you were able to track that down, though. It seems like an easy mistake to make. I'll remember this for any future issues reported on macOS with `fltk`!

> Anyways, I also tried some of the wgpu examples and all give me blank windows on my macbook air. I still haven't had the chance to try things out on my friend's Mac.

You may want to try the `gfx` examples in addition. It might help narrow down whether the problem exists in `wgpu` or in its lower layers. Did you try the `wgpu` examples for 0.6 or 0.7? We have a release forthcoming with 0.7 support, and it would be nice if that fixed some bugs. 😄

MoAlyousef commented:
Thank you. 
I've tried wgpu 0.7. The examples worked well building for release on my Windows machine. On MacOS I haven't had luck. I'll try the gfx examples next. 
I kinda wish I hadn't upgraded to Big Sur, every thing now feels kinda buggy and slow. Link times take a little longer than a minute!

parasyte commented:
> Link times take a little longer than a minute!

I use [`zld`](https://github.com/michaeleisel/zld#if-using-rust) as my default linker. It can still be slow, but is noticeably faster than the native Apple linker. It can be installed with homebrew:

```bash
$ brew tap michaeleisel/zld
$ brew install zld
```

MoAlyousef commented:
Thanks again. I'll check it out
